langserver/fs.go

```diff
@@ -26,42 +26,59 @@ func IsFileSystemRequest(method string) bool {
 		method == "textDocument/didSave"
 }
 
-func (h *HandlerShared) HandleFileSystemRequest(ctx context.Context, req *jsonrpc2.Request) error {
+// HandleFileSystemRequest handles textDocument/did* requests. The path the
+// request is for is returned. true is returned if a file was modified.
+func (h *HandlerShared) HandleFileSystemRequest(ctx context.Context, req *jsonrpc2.Request) (string, bool, error) {
 	span := opentracing.SpanFromContext(ctx)
 	h.Mu.Lock()
 	overlay := h.overlay
 	h.Mu.Unlock()
 
+	var before []byte
+	checkChanged := func(uri string) bool {
+		// We ignore readFile errors, since the []byte being nil gives
+		// us the same comparison.
+		b, _ := h.readFile(ctx, uri)
+		return !bytes.Equal(before, b)
+	}
+
 	switch req.Method {
 	case "textDocument/didOpen":
 		var params lsp.DidOpenTextDocumentParams
 		if err := json.Unmarshal(*req.Params, &params); err != nil {
-			return err
+			return "", false, err
 		}
-		span.SetTag("uri", params.TextDocument.URI)
+		uri := params.TextDocument.URI
+		span.SetTag("uri", uri)
+		before, _ = h.readFile(ctx, uri)
```

Comment:
Is it correct to ignore the error here? What if

1. This call fails (before is now nil)
2. The operation succeeds (in this case didOpen) and modifies the document
3. The h.readFile in checkChanged may also fail and conclude that nil == nil so no change happened?

Think of all the permutations of readFile failing (or not) and the operation modifying the document (or not).

If either of the readFile's fail I don't think we can conclude whether the file changed or not.

Looking at the call site, it seems like it would be prudent to assume that the file changed if either of the readFile calls fail.